---
title: "Mojave_Div_16S_Qiime2"
author: "Nat Pombubpa"
date: "Updated on April 29, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Mojave Endemic Bacteria
This is documentation and codes for Bacterial diversity analysis in Mojave biocrust dataset.

Otu table, taxonomy table, and tree were generated from Qiime2 using DADA2 and Silva database classifier extracted lenght (515-806). Note: Mitochondria and Chloroplast have been filterd and removed from OTU table.

###STEP1: Load all necessary packages for analysis
More information about Phyloseq can be found at the following link: [Phyloseq](https://joey711.github.io/phyloseq/)
If you get error in this step, you probably need to install any packages which causes error.

```{r warning=FALSE, message=FALSE}
library(ape)
library(vegan)
library(plyr)
library(dplyr)
library(scales)
library(grid)
library(reshape2)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(ggpubr)
library(data.table)
library(tidyr)
library(tidyverse)
library(multcompView)
library(VennDiagram)
```

####Function to summarize taxonomy 
######plot_taxa_summary(GlobalPatterns, "Phylum")
######plot_taxa_summary(GlobalPatterns, "Phylum", "SampleType")
######summarize_taxa(GlobalPatterns, "Phylum")
######summarize_taxa(GlobalPatterns, "Phylum", "SampleType")

######This function was written by Joey (but it is not part of phyloseq package)
######[https://github.com/joey711/phyloseq/issues/418](https://github.com/joey711/phyloseq/issues/418)

```{r warning=FALSE}
fast_melt = function(physeq){
  # supports "naked" otu_table as `physeq` input.
  otutab = as(otu_table(physeq), "matrix")
  if(!taxa_are_rows(physeq)){otutab <- t(otutab)}
  otudt = data.table(otutab, keep.rownames = TRUE)
  setnames(otudt, "rn", "taxaID")
  # Enforce character taxaID key
  otudt[, taxaIDchar := as.character(taxaID)]
  otudt[, taxaID := NULL]
  setnames(otudt, "taxaIDchar", "taxaID")
  # Melt count table
  mdt = melt.data.table(otudt, 
                        id.vars = "taxaID",
                        variable.name = "SampleID",
                        value.name = "count")
  # Remove zeroes, NAs
  mdt <- mdt[count > 0][!is.na(count)]
  # Calculate relative abundance
  mdt[, RelativeAbundance := count / sum(count), by = SampleID]
  if(!is.null(tax_table(physeq, errorIfNULL = FALSE))){
    # If there is a tax_table, join with it. Otherwise, skip this join.
    taxdt = data.table(as(tax_table(physeq, errorIfNULL = TRUE), "matrix"), keep.rownames = TRUE)
    setnames(taxdt, "rn", "taxaID")
    # Enforce character taxaID key
    taxdt[, taxaIDchar := as.character(taxaID)]
    taxdt[, taxaID := NULL]
    setnames(taxdt, "taxaIDchar", "taxaID")
    # Join with tax table
    setkey(taxdt, "taxaID")
    setkey(mdt, "taxaID")
    mdt <- taxdt[mdt]
  }
  return(mdt)
}

summarize_taxa = function(physeq, Rank, GroupBy = NULL){
  Rank <- Rank[1]
  if(!Rank %in% rank_names(physeq)){
    message("The argument to `Rank` was:\n", Rank,
            "\nBut it was not found among taxonomic ranks:\n",
            paste0(rank_names(physeq), collapse = ", "), "\n",
            "Please check the list shown above and try again.")
  }
  if(!is.null(GroupBy)){
    GroupBy <- GroupBy[1]
    if(!GroupBy %in% sample_variables(physeq)){
      message("The argument to `GroupBy` was:\n", GroupBy,
              "\nBut it was not found among sample variables:\n",
              paste0(sample_variables(physeq), collapse = ", "), "\n",
              "Please check the list shown above and try again.")
    }
  }
  # Start with fast melt
  mdt = fast_melt(physeq)
  if(!is.null(GroupBy)){
    # Add the variable indicated in `GroupBy`, if provided.
    sdt = data.table(SampleID = sample_names(physeq),
                     var1 = get_variable(physeq, GroupBy))
    setnames(sdt, "var1", GroupBy)
    # Join
    setkey(sdt, SampleID)
    setkey(mdt, SampleID)
    mdt <- sdt[mdt]
  }
  # Summarize
  Nsamples = nsamples(physeq)
  summarydt = mdt[, list(meanRA = sum(RelativeAbundance)/Nsamples,
                         sdRA = sd(RelativeAbundance),
                         minRA = min(RelativeAbundance),
                         maxRA = max(RelativeAbundance)),
                  by = c(Rank, GroupBy)]
  return(summarydt)
}

plot_taxa_summary = function(physeq, Rank, GroupBy = NULL){
  # Get taxa summary table 
  dt1 = summarize_taxa(physeq, Rank = Rank, GroupBy = GroupBy)
  # Set factor appropriately for plotting
  RankCol = which(colnames(dt1) == Rank)
  setorder(dt1, -meanRA)
  dt1[, RankFac := factor(dt1[[Rank]], 
                          levels = rev(dt1[[Rank]]))]
  dt1[, ebarMax := max(c(0, min(meanRA + sdRA))), by = eval(Rank)]
  dt1[, ebarMin := max(c(0, min(meanRA - sdRA))), by = eval(Rank)]
  # Set zeroes to one-tenth the smallest value
  ebarMinFloor = dt1[(ebarMin > 0), min(ebarMin)]
  ebarMinFloor <- ebarMinFloor / 10
  dt1[(ebarMin == 0), ebarMin := ebarMinFloor]

  pRank = ggplot(dt1, aes(x = meanRA, y = RankFac)) +
    scale_x_log10() +
    xlab("Mean Relative Abundance") +
    ylab(Rank) +
    theme_bw()
  if(!is.null(GroupBy)){
    # pRank <- pRank + facet_wrap(facets = as.formula(paste("~", GroupBy)))
    pRank <- pRank + geom_point(mapping = aes_string(colour = GroupBy),
                                size = 5)
  } else {
    # Don't include error bars for faceted version
    pRank <- pRank + geom_errorbarh(aes(xmax = ebarMax,
                                        xmin = ebarMin))
  }
  return(pRank)
}


```

###STEP2: Transition data from Qiime2 to Phyloseq
Note: otu table and taxonomy table have been fixed right after Qiime2. Nat's Qiime2 script already include this step.
However, you can simply fix by changing "#Feature ID" to "OTU ID" ot "OTU.ID" and the following commnand should work fine.

####STEP2.1: Import otu table from Qiime2

```{r}
otu = read.table(file="MojaveBacQ2FWLEsilva515806.otu_table.fix.txt", header=T, sep='\t')
```

```{r}
head(otu)
```

Please note dimension of otu table. For this project, there are 21697 rows(OTU.ID) which will need to be matched to taxonomy table before starting Phyloseq analysis.
```{r}
dim(otu)
```

####STEP2.2: Import taxonomy table from Qiime2

```{r}
tax <- read.table(file="taxonomy.fix.tsv", sep='\t', header=TRUE)
head(tax)
```

Taxonomy table contains 22081 rows(OTU.ID) which is more than otu table. (Mitochondria and Chloroplast were removed in otu table which resulted in less number of OTU.ID in otu table.)
```{r}
dim(tax)
```

####STEP2.3: Match OTU.ID between otu and taxonomy table
Keep only OTU.ID in taxonomy table  that were found in otu table.
```{r}
tax_filtered <- tax[tax$OTU.ID %in% otu$OTU.ID,]
head(tax_filtered)
```

Check dimension of tax_filterd table: the number of OTU.ID should be the same as OTU table (in this case: 21697 OTU.ID)
```{r}
dim(tax_filtered)
```

Number of rows for "tax_filtered" and "otu" should be the same.
```{r}
dim(otu)
```

####STEP2.4: Select only OTU.ID column and taxonomy for taxonomy table
Taxonomy table from Qiime2 also contains the third column which is "Confidence". We will remove this column before loading data to Phyloseq.
```{r}
tax_filtered = tax_filtered[,c(1,2)]
```

After selection, taxonomy column will be separated into 7 levels as "Kingdom","Phylum","Class","Order", "Family", "Genus","Species"
```{r}
tax_filtered = separate(tax_filtered, Taxon, c("Kingdom","Phylum","Class","Order", "Family", "Genus","Species"), sep= ";", remove=TRUE)
```

At the end of this step, taxonomy table will contain OTU.ID column + 7 levels taxonomy columns (total of 8 columns). 
```{r}
dim(tax_filtered)
```

```{r}
head(tax_filtered)
```

Save "filtered" taxonomy table for Phyloseq and future usage.
```{r}
write.csv(tax_filtered, file="taxonomy_phyloseq_fix.csv")
```

###STEP3: Import otu table for analysis using Phyloseq 

```{r warning=FALSE}
otus <- read.table("MojaveBacQ2FWLEsilva515806.otu_table.fix.txt",
                   header=T,sep="\t",row.names = 1)
otumat <- as(as.matrix(otus), "matrix")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

###STEP4: Import taxonomy table for analysis using Phyloseq
This taxonomy table was created from STEP2.
```{r warning=FALSE}
taxmat <- read.csv("taxonomy_phyloseq_fix.csv", 
                   sep=",",row.names=1)
row.names(taxmat) = taxmat$OTU.ID
taxmat <- as(as.matrix(taxmat),"matrix")
TAX = tax_table(taxmat)
```

###STEP5: Import tree file for analysis using Phyloseq

```{r warning=FALSE}
treefile = "tree.nwk"
tree = read.tree(treefile)
```

###STEP6: Import mapping file for analysis using Phyloseq

1.Check mapping file before import to R, R will automatically change sample name that starts with number or contain “-” in sample name. If you get error in this step, you should check sample name first. 

2.First column of first row should not start with #, R will not read the first row that starts with #

3. You can choose which samples to include in analysis by indicating specific group in the column

```{r warning=FALSE}
meta = read.table("Mojave_mappingfile_8-Aug-2018.txt",
                  header=TRUE,row.names=1,
                  sep="\t",stringsAsFactors=FALSE)
#meta <- meta[which(meta$Layer %in% c("Surface")),]
```

Check if your metadata file has been import successfully and correctly, the output will show a table of your metadata file (mapping file). *If you do not have header, you might start your first row with # (remove # and reload your mapping file).

```{r}
head(meta)
```

Construct sample_data-class using imported metadata
```{r warning=FALSE}
sampleData <- sample_data(meta)
```

###STEP7: Construct Phyloseq object
To construct phyloseq object, otu table, taxonomy table, and sampleData are required. Phylogenetic tree can be included, but it is not necessary for constructing phyloseq object.
Construct Phyloseq object called "Physeq"

```{r warning=FALSE}
physeq = phyloseq(OTU,TAX,sampleData,tree)
```

###STEP8: Check phyloseq object
This should indicate that your physeq is a "phyloseq-class experiment-level object"

```{r}
physeq
```

###STEP9: Remove ASVs
Remove any ASVs that present only 5 time.

```{r}
physeq.prune = prune_taxa(taxa_sums(physeq) > 5, physeq)
```

```{r}
physeq.prune
```

###STEP10: Plot read counts to check dataset
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

```{r}
readcount = data.table(as(sample_data(physeq.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")
#For plotting, use command below.
#SeqDepth = ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
```

TotalReads of all the samples can be in this table (select only SampleID and TotalReads columns).
In order to check samples with low number of reads, "order()" can be used to sort "TotalReads" column.
In this dataset, N55.Rhizo has very low number of reads, so will will filter this sample out using the next minimum number of reads.
```{r}
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
```

```{r}
head(readcount)
```

###STEP11: Rarefy OTUs to a minimum number of reads
Rarefy OTUs (remove any samples that has very low number of reads)
If this step work fine, you have successfully imported data to R and completely generate phyloseq object as "physeq.prune.rarefy".
```{r warning=FALSE}
set.seed(711)
physeq.prune.rarefy = rarefy_even_depth(physeq.prune, sample.size = 37435, replace = FALSE, trimOTUs = TRUE)
physeq.prune.rarefy
```

###STEP12.1: Plot Alpha diversity by Crust_type
Alpha diversity can be Chao1, Observed, Shannon, Simpson
This plot include statistical analysis using "stat_compare_means" with "method = anova"

###Reorder Crust_type
```{r}
#physeq.prune.rarefy.SF = subset_samples(physeq.prune.rarefy, Layer=="Surface")
sample_data(physeq.prune.rarefy)$Crust_type = factor(sample_data(physeq.prune.rarefy)$Crust_type, levels = c("LAC","CLC", "GLC", "RMC", "SMC","LAC.Sub","CLC.Sub", "GLC.Sub", "RMC.Sub", "SMC.Sub"))
```

###ANOVA
```{r}
alpha.diversity = estimate_richness(physeq.prune.rarefy, measures = c("Observed"))
data.anova = cbind(sample_data(physeq.prune.rarefy), alpha.diversity)
physeq.prune.rarefy.anova = aov(Observed ~ Crust_type, data.anova)
summary(physeq.prune.rarefy.anova)
```

```{r}
tukey.ps = TukeyHSD(x=physeq.prune.rarefy.anova, 'Crust_type', conf.level = 0.95)
```

```{r}
plot(tukey.ps , las=1 , col="brown" )
```


```{r}
generate_label_df <- function(tukey.ps, variable){
 
     # Extract labels and factor levels from Tukey post-hoc 
     Tukey.levels <- tukey.ps[[variable]][,4]
     Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])
     
     #I need to put the labels in the same order as in the boxplot :
     Tukey.labels$treatment=rownames(Tukey.labels)
     Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
     return(Tukey.labels)
     }
```

```{r}
LABELS=generate_label_df(tukey.ps, 'Crust_type')
```

```{r}
names(LABELS) = c('Letters','Crust_type')
```

```{r}
LABELS
```


```{r}
ylabel <- data.frame("ylabel" = c(1150,2100,1050,2050,850,2650,1200,2500,1100,1900))
```

```{r}
LABELS$ylabel<-ylabel$ylabel
```

```{r}
#physeq.prune.rarefy.plot.richness.Crust_type = plot_richness(physeq.prune.rarefy, x="Crust_type", color=("Crust_type"), measures=c("Observed")) + geom_boxplot(lwd=1,fill=c("#999999", "gold2", "cadetblue3", "palegreen2", "lightpink")) + ggtitle("Bacterial Alpha Diversity by Crust_type") + stat_compare_means(method = "anova", label.y = 1222) + theme(plot.title = element_text(hjust = 0.5)) + scale_colour_manual(values = c("#000000", "#E69F00", "#56B4E9", "#009E73", "#CC79A7")) + geom_text(data=LABELS, aes(x=LABELS$Crust_type, y = 1150, label = LABELS$Letters))

#physeq.prune.rarefy.plot.richness.Crust_type 
```

```{r}
physeq.prune.rarefy.plot.richness.Crust_type = plot_richness(physeq.prune.rarefy, x="Crust_type", color=("Crust_type"), measures=c("Observed")) + geom_boxplot() + ggtitle("Mojave Biocrust Bacterial Alpha Diversity by Crust type and Layer") + stat_compare_means(method = "anova", label.y = 2800) + theme(plot.title = element_text(hjust = 0.5)) + geom_text(data=LABELS, aes(x=LABELS$Crust_type, y = LABELS$ylabel , label = LABELS$Letters))

physeq.prune.rarefy.plot.richness.Crust_type
```


```{r warning=FALSE}
pdf("./Figures/Mojave Biocrust Bacterial Alpha Diversity by Crust type and Layer.pdf", width = 8, height = 5 )
physeq.prune.rarefy.plot.richness.Crust_type
dev.off()
```

```{r}
physeq.prune.rarefy.plot.richness.Crust_type = plot_richness(physeq.prune.rarefy, x="Crust_type", color=("Crust_type"), measures=c("Observed")) + geom_boxplot() + ggtitle("Mojave Biocrust Bacterial Alpha Diversity (Observed) by Crust type") + stat_compare_means(method = "anova", label.y = 3000) + stat_compare_means(aes(label=..p.signif..), method = "t.test", ref.group = "LAC", label.y = 2800, hide.ns = TRUE) + theme_bw() + theme(plot.title = element_text(hjust = 0.5))

physeq.prune.rarefy.plot.richness.Crust_type
```

To save Alpha diversity (Observed) by Crust_type plot as a pdf into Figures directory.

```{r warning=FALSE}
pdf("./Figures/Mojave Biocrust Bacterial Alpha (Observed) Diversity by Crust type.pdf", width = 8, height = 5 )
physeq.prune.rarefy.plot.richness.Crust_type
dev.off()
```

###Reorder Site and layer
```{r}
sample_data(physeq.prune.rarefy)$Site_layer = factor(sample_data(physeq.prune.rarefy)$Site_layer, levels = c("JTNP_SF","CIMA_SF", "GMT_SF", "KELSO_SF", "JTNP_Sub","CIMA_Sub", "GMT_Sub", "KELSO_Sub"))
```

###Reorder Site
```{r}
#physeq.prune.rarefy.SF = subset_samples(physeq.prune.rarefy, Layer=="Surface")
#sample_data(physeq.prune.rarefy.SF)$Site = factor(sample_data(physeq.prune.rarefy.SF)$Site, levels = c("JTNP","CIMA", "GMT", "KELSO"))
```


###ANOVA
```{r}
alpha.diversity = estimate_richness(physeq.prune.rarefy, measures = c("Observed"))
data.anova = cbind(sample_data(physeq.prune.rarefy), alpha.diversity)
physeq.prune.rarefy.anova = aov(Observed ~ Site_layer, data.anova)
summary(physeq.prune.rarefy.anova)
```

```{r}
physeq.prune.rarefy.anova
```

```{r}
tukey.ps = TukeyHSD(x=physeq.prune.rarefy.anova, 'Site_layer', conf.level = 0.95)
```

```{r}
generate_label_df <- function(tukey.ps, variable){
 
     # Extract labels and factor levels from Tukey post-hoc 
     Tukey.levels <- tukey.ps[[variable]][,4]
     Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])
     
     #I need to put the labels in the same order as in the boxplot :
     Tukey.labels$treatment=rownames(Tukey.labels)
     Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
     return(Tukey.labels)
     }
```


```{r}
LABELS=generate_label_df(tukey.ps, 'Site_layer')
```

```{r}
names(LABELS) = c('Letters','Site_layer')
```

```{r}
ylabel <- data.frame("ylabel" = c(1150,2650,1200,2050,1000,1650,1100,2500))
```

```{r}
LABELS$ylabel<-ylabel$ylabel
```


```{r}
LABELS
```

```{r}
#physeq.prune.rarefy.plot.richness.Site_layer = plot_richness(physeq.prune.rarefy.SF, x="Site", color=("Site"), measures=c("Observed")) + geom_boxplot(lwd=1,fill=c("#999999", "gold2", "cadetblue3", "palegreen2")) + ggtitle("Bacterial Alpha Diversity by Site") + stat_compare_means(method = "anova", label.y = 1222) + theme(plot.title = element_text(hjust = 0.5)) + scale_colour_manual(values = c("#000000", "#E69F00", "#56B4E9", "#009E73")) + geom_text(data=LABELS, aes(x=LABELS$Site, y = 1150, label = LABELS$Letters))

#physeq.prune.rarefy.plot.richness.Site_layer 
```

```{r}
physeq.prune.rarefy.plot.richness.Site_layer = plot_richness(physeq.prune.rarefy, x="Site_layer", color=("Site_layer"), measures=c("Observed")) + geom_boxplot() + ggtitle("Mojave Biocrust Bacterial Alpha Diversity by Site and Layer") + stat_compare_means(method = "anova", label.y = 2800) + theme(plot.title = element_text(hjust = 0.5)) + geom_text(data=LABELS, aes(x=LABELS$Site, y = LABELS$ylabel , label = LABELS$Letters))

physeq.prune.rarefy.plot.richness.Site_layer
```

```{r warning=FALSE}
pdf("./Figures/Mojave Biocrust Bacterial Alpha Diversity by Site and Layer.pdf", width = 8, height = 5 )
physeq.prune.rarefy.plot.richness.Site_layer
dev.off()
```

```{r}
physeq.prune.rarefy.plot.richness.Site_layer = plot_richness(physeq.prune.rarefy, x="Site_layer", color=("Site_layer"), measures=c("Observed")) + geom_boxplot() + ggtitle("Mojave Biocrust Bacterial Alpha Diversity (Observed) by Site and Layer") + stat_compare_means(method = "anova", label.y = 3000) + stat_compare_means(aes(label=..p.signif..), method = "t.test", ref.group = "JTNP_SF", label.y = 2800, hide.ns = TRUE) + theme_bw() + theme(plot.title = element_text(hjust = 0.5))

physeq.prune.rarefy.plot.richness.Site_layer
```

To save Alpha diversity (Observed) by Site and Layer plot as a pdf into Figures directory.

```{r warning=FALSE}
pdf("./Figures/Mojave Biocrust Bacterial Alpha (Observed) Diversity by Site and Layer.pdf", width = 8, height = 5 )
physeq.prune.rarefy.plot.richness.Site_layer
dev.off()
```

###Rarefaction curve
extract OTU table, transpose and convert to data frame
This step wil take several minutes.
```{r}
otu.rare <- otu_table(physeq.prune.rarefy)
otu.rare <- as.data.frame(t(otu.rare))
sample_names <- rownames(otu.rare)

#vegan rarecurve 
otu.rarecurve <- rarecurve(otu.rare, step = 1000, sample = 37435, label = T)
```

```{r}
rare <- lapply(otu.rarecurve, function(x){
  b <- as.data.frame(x)
  b <- data.frame(ASV = b[,1], Sample.size = rownames(b))
  b$Sample.size <- as.numeric(gsub("N", "",  b$Sample.size))
  return(b)
})
```
label list
```{r}
names(rare) <- sample_names
```
convert to data frame
```{r}
#tidyverse
rare <- map_dfr(rare, function(x){
  z <- data.frame(x)
  return(z)
}, .id = "sample")
```

```{r}
head(rare)
```

```{r}
ggplot(data = rare)+
  geom_line(aes(x = Sample.size, y = ASV, color = sample), show.legend = FALSE)+
  scale_x_continuous(labels =  waiver()) + theme_bw() + ggtitle("Bacterial Rarefaction Curve")+
  theme(plot.title = element_text(hjust = 0.5))
```

Save rarefaction curve as pdf.
```{r}
pdf("./Figures/Bacterial Rarefaction Curve.pdf", width = 8, height = 5 )
ggplot(data = rare)+
  geom_line(aes(x = Sample.size, y = ASV, color = sample), show.legend = FALSE)+
  scale_x_continuous(labels =  waiver()) + theme_bw() + ggtitle("Bacterial Rarefaction Curve")+
  theme(plot.title = element_text(hjust = 0.5))
dev.off()
```

###STEP12.2 Taxonomic composition
```{r fig.height=5, fig.width=8, fig.align="center"}
taxcom_sitelayer = ggplot(data = psmelt(physeq.prune.rarefy), mapping = aes_string(x = "Site_layer" ,y = "Abundance", fill = "Phylum" )) + 
  geom_bar(stat="identity", position="fill") + 
  ggtitle("Bacterial Taxonomic Composition by Site and Layer")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))+
  theme(plot.title = element_text(hjust = 0.5))

#print(taxcom_sitelayer)
```

```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Bacterial Taxonomic Composition by Site and Layer.pdf", width = 10, height = 6 )
taxcom_sitelayer
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}
taxcom_site = ggplot(data = psmelt(physeq.prune.rarefy.SF), mapping = aes_string(x = "Site" ,y = "Abundance", fill = "Phylum" )) + 
  geom_bar(stat="identity", position="fill") + 
  ggtitle("Bacterial Taxonomic Composition by Site")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))+
  theme(plot.title = element_text(hjust = 0.5))

#print(taxcom_sitelayer)
```

```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Bacterial Taxonomic Composition by Site.pdf", width = 10, height = 6 )
taxcom_site
dev.off()
```


```{r fig.height=5, fig.width=8, fig.align="center"}
taxcom_crusttype = ggplot(data = psmelt(physeq.prune.rarefy), mapping = aes_string(x = "Crust_type" ,y = "Abundance", fill = "Phylum" )) + 
  geom_bar(stat="identity", position="fill") + 
  ggtitle("Bacterial Taxonomic Composition by Crust type")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))+
  theme(plot.title = element_text(hjust = 0.5))

#print(taxcom_crusttype)
```

```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Bacterial Taxonomic Composition by Crust type.pdf", width = 10, height = 6 )
taxcom_crusttype
dev.off()
```


###STEP12.3 Comparing Taxonomic Differences
```{r}
BacPhylum = as.character(get_taxa_unique(physeq.prune.rarefy.SF, "Phylum"))
BacPhylum = BacPhylum[complete.cases(BacPhylum)]
BacPhylum
```

Comparing by Site and Layer
```{r}
pdf("./Figures/Bacterial Phylum Alpha (Observed) Diversity by Site.pdf", width = 8, height = 5 )
for (i in BacPhylum) {
  APS = subset_taxa(physeq.prune.rarefy.SF, Phylum == i)
  APS.plot = plot_richness(APS, x="Site", color =("Site"), measures=c("Observed")) + geom_boxplot(lwd=1,fill=c("#999999", "gold2", "cadetblue3", "palegreen2")) + ggtitle(i) + stat_compare_means(method = "anova") + theme(plot.title = element_text(hjust = 0.5)) + scale_colour_manual(values = c("#000000", "#E69F00", "#56B4E9", "#009E73"))
  print(APS.plot)
}
dev.off()
```


```{r}
physeq.prune.rarefy.SF.Acido = subset_taxa(physeq.prune.rarefy.SF, Phylum == "D_1__Firmicutes")
alpha.diversity = estimate_richness(physeq.prune.rarefy.SF.Acido, measures = c("Observed"))
data.anova = cbind(sample_data(physeq.prune.rarefy.SF.Acido), alpha.diversity)
physeq.prune.rarefy.SF.Acido.anova = aov(Observed ~ Site, data.anova)
tukey.ps.SF = TukeyHSD(x=physeq.prune.rarefy.SF.Acido.anova, 'Site', conf.level = 0.95)
LABELS=generate_label_df(tukey.ps.SF, 'Site')
names(LABELS) = c('Letters','Site')
```

```{r}
physeq.prune.rarefy.plot.richness.Site.Acido = plot_richness(physeq.prune.rarefy.SF.Acido, x="Site", color=("Site"), measures=c("Observed")) + geom_boxplot(lwd=1,fill=c("#999999", "gold2", "cadetblue3", "palegreen2")) + ggtitle("Firmicutes Alpha Diversity by Site") + stat_compare_means(method = "anova", label.y = 85) + theme(plot.title = element_text(hjust = 0.5)) + scale_colour_manual(values = c("#000000", "#E69F00", "#56B4E9", "#009E73")) + geom_text(data=LABELS, aes(x=LABELS$Site, y = 78, label = LABELS$Letters))

physeq.prune.rarefy.plot.richness.Site.Acido
```

```{r}
APS.plot = plot_richness(APS, x="Site", color =("Site"), measures=c("Observed")) + geom_boxplot() + ggtitle(i) + stat_compare_means(method = "anova") + theme(plot.title = element_text(hjust = 0.5))
```




Comparing by Crust type
```{r}
pdf("./Figures/Bacterial Phylum Alpha (Observed) Diversity by Crust_type .pdf", width = 8, height = 5 )
for (i in BacPhylum) {
  APS = subset_taxa(physeq.prune.rarefy, Phylum == i)
  APS.plot = plot_richness(APS, x="Crust_type", color =("Crust_type"), measures=c("Observed")) + geom_boxplot() + ggtitle(i) + stat_compare_means(method = "anova") + theme(plot.title = element_text(hjust = 0.5))
  print(APS.plot)
}
dev.off()
```


###STEP12.4 Summarize taxa composition

```{r}
taxa.summary.by.phylum = summarize_taxa(physeq.prune.rarefy, "Phylum")
```

######To save taxa summary into text file under your current working directory.

```{r warning=FALSE}
write.table(taxa.summary.by.phylum, file = "taxa.summary.by.phylum.txt", sep ="\t")
```




###Venn Diagram

```{r}
venn_diagram2 <- function(a, b, name_a = "A", name_b = "B", colors =  c("#e41a1c","#377eb8"), out = FALSE, file_name = "default.pdf", euler = FALSE, main = "main"){
	# Niels Hanson
  # https://github.com/hallamlab/mp_tutorial/blob/master/downstream_analysis_r/code/venn_diagram2.r
	# Info: A function to create a venn digagram between three (3) overapping sets a, b, and c.
	# Requies: The VennDiagram R package, the program will try to download and install if you dont
	#          have it.
	# Inputs: a,b,c - vertical vectors or the first column of matrixes as sets of strings to be compared
	#         name_a, name_b, and name_c - names that you give to sets a, b, and c, respectively
	#         these will appear in the final image
	#         colors - a vector of colors defined in hex format ie. colors <- c("#B80830","#EACC33","#46E2D9")
	#         it might be a good idea that these mix well as the interlapping classes will be a mix of colors
	#         Colors must be specified in hex ie. "#123456", google "hex colors" for more info
	#         name_output - the name of the output image that will be put in R's current working directory
	#         you can set this yourself before running. ie. setwd("~/my-director/")
	
	# A: find all classes for set a, a_b, a_c, a_b_c
	# number in a, we will use this at the end for a sanity check
	
	# make sure that a,b,c are all sets, get rid of duplicates
	a <- intersect(a,a)
	b <- intersect(b,b)
	# length of a
	a_tot <- length(a)
	# length of b
	b_tot <- length(b)
	# number shared between a and b and c
	a_b <- intersect(a,b)
	a_b_tot <- length(a_b)
	a_b_tot
	a_only <- setdiff(a,b)
	a_only_tot <- length(a_only)
	b_only <- setdiff(b,a)
	b_only_tot <- length(b_only)

	# try to load VennDiagram else try to download it
	# More info: see Hanbo Chen and Paul C Boutros. BMC Bioinformatics 2011, 12:35 doi:10.1186/1471-2105-12-35
	try(library(VennDiagram), install.packages("VennDiagram")) 
	library(VennDiagram)
	

	# create the venn diagram and output into the current working directory
	temp <- list(
		name_a = c(1:length(a)),
		name_b = c((length(a_only)+1):((length(a_only))+length(b)))
		)
	names(temp) <- c(name_a,name_b)
	
	output <- venn.diagram(
		x = temp,
		filename = NULL,
		col = "grey28",
		fill = c(colors[1], colors[2]),
		alpha = 0.58,
		label.col = c("grey28"),
		cex = 2.5,
		fontface = "bold",
		fontfamily = "sans",
		cat.default.pos = "text",
		cat.col = c("grey28"),
		cat.cex = 2.5,
		cat.fontfamily = "sans",
		cat.dist = c(0.06, 0.06),
		cat.pos = c(-20, 14),
		euler.d = euler,
		scaled = euler,
		main = main,
		main.fontface = "bold",
		main.fontfamily = "sans"
		);
		
		try(library("grid"), install.packages("grid")) 
    library(grid)
        if (out == TRUE) {
        # only produce ouptut files upon request
        png(filename = paste(file_name,'.png', sep=''))
        grid.draw(output)
        dev.off()       
        pdf(file = paste(file_name,'.pdf', sep=''), width=8, height=8)
        grid.draw(output)
        dev.off()
        } else {
          # just plot the figure
          grid.draw(output)
       }
		
        # pack all the sets up and return to the user
        out <- NULL;
         a <- c(out, name_a, as.vector(a))
         a_only <-  as.vector(a_only)
         a_b <- as.vector(a_b)
         b <- c(out,name_b,as.vector(b))
         b_only <- c(out,paste(name_b,"only",sep="_"),as.vector(b_only))
        out <- list(a,a_only,a_b,b,b_only)
        names(out) <- c(name_a, 
                        paste(name_a,"only",sep="_"), 
                        paste(name_a,name_b,sep="_"),
                        name_b,
                        paste(name_b,"only",sep="_"))
  
        return(out)

}
```

```{r}
Bac.physeq.prune.rarefy = subset_taxa(physeq.prune.rarefy, Kingdom != "Rhizaria")
Bac.physeq.prune.rarefy = subset_taxa(Bac.physeq.prune.rarefy, Kingdom != "Chromista")
Bac.physeq.prune.rarefy = subset_taxa(Bac.physeq.prune.rarefy, Kingdom != "Unassigned")
Bac.physeq.prune.rarefy = subset_taxa(Bac.physeq.prune.rarefy, Kingdom != "D_0__Eukaryota")
Bac.physeq.prune.rarefy
```

###Subset for Surface and Subsurface
```{r}
Bac.physeq.prune.rarefy.SF = subset_samples(Bac.physeq.prune.rarefy, Layer=="Surface")
Bac.physeq.prune.rarefy.SF = prune_taxa(taxa_sums(Bac.physeq.prune.rarefy.SF) > 5, Bac.physeq.prune.rarefy.SF)
Bac.physeq.prune.rarefy.SF
```

```{r}
Bac.physeq.prune.rarefy.SUB = subset_samples(Bac.physeq.prune.rarefy, Layer=="Subsurface")
Bac.physeq.prune.rarefy.SUB = prune_taxa(taxa_sums(Bac.physeq.prune.rarefy.SUB) > 5, Bac.physeq.prune.rarefy.SUB)
Bac.physeq.prune.rarefy.SUB
```

```{r}
OTU.name.Bac.physeq.prune.rarefy.SF = rownames((otu_table(Bac.physeq.prune.rarefy.SF)))
```

```{r}
OTU.name.Bac.physeq.prune.rarefy.SUB = rownames((otu_table(Bac.physeq.prune.rarefy.SUB)))
```

```{r}
## plot venn diagrams
      
# Surface vs Subsurface
pdf("./Figures/Bacteria_Surface_Subsurface_venn.pdf", width = 8, height = 8)
ven_test = venn_diagram2(OTU.name.Bac.physeq.prune.rarefy.SF, OTU.name.Bac.physeq.prune.rarefy.SUB, "Surface", "Subsurface", colors= c("#009E73", "#CC79A7"), euler=TRUE, main = "Bacteria Surface-Subsurface Venn")
dev.off()
```

###List of all core OTUs
```{r}
ven_test$Surface_Subsurface
```




```{r}
str(a)
```

