---
title: "Mojave_Endemic_ITS_phyloseq"
author: "Nat Pombubpa"
date: "Updated on April 29, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###STEP1: Load all necessary packages for analysis
More information about Phyloseq can be found at the following link: [Phyloseq](https://joey711.github.io/phyloseq/)
If you get error in this step, you probably need to install any packages which causes error.

```{r warning=FALSE, message=FALSE}
library(ape)
library(vegan)
library(dplyr)
library(scales)
library(grid)
library(reshape2)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(ggpubr)
library(plyr)
library(data.table)
library(tidyr)
library(tidyverse)
```

###STEP2: Import Mapping file (metadate file)
1.Check mapping file before import to R, R doesn’t seem to like sample name to start with number or contain “-” in sample name. If you get error in this step, you should check file name first.

2.First column of first row should not start with #, R will not read the first row that starts with #

3. You can choose which samples to include in analysis by indicating "KEEP" in Description column

```{r}
meta = read.table("Mojave_mappingfile_8-Aug-2018.txt",header=TRUE,row.names=1,sep="\t",stringsAsFactors=FALSE)
```

If need to keep or remove some samples
```{r}
meta <- meta[which(meta$Layer %in% c("Surface")),]
```

###STEP3: Check if your metadata file has been import successfully and correctly

The output will show a table of your metadata file (mapping file).

*If you do not have header, you might start your first row with #

```{r warning=FALSE}
head(meta)
```

###STEP4: Construct sample_data-class using imported metadata

```{r}
sampleData <- sample_data(meta)
```

###STEP5: Import OTU table

OTU table from Mojave 16S data is “MC2017FC.otu_table.txt”.
MC-Mojave Crust-2017-FC-Fungi completed

```{r}
otus <- read.table("MC2018FC.otu_table.txt",header=T,sep="\t",row.names=1)
otumat <- as(as.matrix(otus), "matrix")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

Check imported OTU table

```{r warning=FALSE}
head(OTU)
```

###STEP6: Import taxonomy table
Taxonmy table generated from AMPtk need to be rearranged using following script.

“perl rdp_taxonmy2mat.pl<Input_taxonmy.txt>Output_taxonomy.txt”

rdp_taxonomy2mat.pl was created by Professor Jason E. Stajich

```{r}
taxmat <- read.table("MC2018FC.taxonomy.fix.txt", header=T,sep="\t",row.names=1)
taxmat <- as(as.matrix(taxmat),"matrix")
TAX = tax_table(taxmat)
```

###STEP7: Import phylogenetic tree
Phylogenetic tree can also be include for further phylogenetic analysis.

```{r warning=FALSE}
treefile = "MC2018FC.tree.phy"
tree = read.tree(treefile)
```

###STEP8: Construct Phyloseq object
To construct phyloseq object, otu table, taxonomy table, and sampleData are required. Phylogenetic tree can be included, but it is not necessary for constructing phyloseq object.
Construct Phyloseq object called "Physeq"

```{r warning=FALSE}
physeq = phyloseq(OTU,TAX,sampleData,tree)
```

Check phyloseq object
This should indicate that your physeq is a "phyloseq-class experiment-level object""

```{r warning=FALSE}
physeq
```

###STEP9: Remove singletons
Remove any OTUs that present only one time.

```{r }
physeq.prune = prune_taxa(taxa_sums(physeq) > 1, physeq)
```

```{r warning=FALSE}
physeq.prune
```

###STEP10: Plot read counts to check dataset
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

```{r}
readcount = data.table(as(sample_data(physeq.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")

#For plotting, use command below.
#SeqDepth = ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
```

TotalReads of all the samples can be in this table (select only SampleID and TotalReads columns).
In order to check samples with low number of reads, "order()" can be used to sort "TotalReads" column.
In this dataset, N55.Rhizo has very low number of reads, so will will filter this sample out using the next minimum number of reads.
```{r}
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
```

```{r}
head(readcount)
```

```{r warning=FALSE}
set.seed(1)
physeq.prune.rarefy = rarefy_even_depth(physeq.prune, sample.size = 31107, replace = FALSE, trimOTUs = FALSE)
physeq.prune.rarefy
```

###STEP12:Merge samples by metadata 
For endemic analysis, we would like to look for endemic species at each sites. Therefore, we will merge samples from each site together. For examples, all of samples from GMT will be merged into one group as GMT. After merging, there will be 4 groups which are GMT, KELSO, CIMA, and JTNP.

```{r}
physeq.prune.rarefy.merge.sitelayer <- merge_samples(physeq.prune.rarefy, "Site")
sample_data(physeq.prune.rarefy.merge.sitelayer)$Site_layer <- factor(sample_names(physeq.prune.rarefy.merge.sitelayer))
physeq.prune.rarefy.merge.sitelayer
```

###STEP14:Prep data for endemic species filtering
Endemic species filtering will be done by transform abundance data into presence/absence data. Then, the data will be selected if the sum of each row is equal to 1 which means that specific OTU (row) only have one presence (one site).

####STEP14.1:Transform data to presence/absence
```{r}
physeq.prune.rarefy.merge.sitelayer.binary = transform_sample_counts(physeq.prune.rarefy.merge.sitelayer, function(abund) 1*(abund>0))
```

####STEP14.2:Extract OTU(ASV) table for filtering
Extract OTU tabel from phyloseq object and transpose the table so that row names are OTU ID.
```{r}
OTU_SF_merge_site = otu_table(physeq.prune.rarefy.merge.sitelayer.binary)
OTU_SF_merge_site = t(OTU_SF_merge_site)
head(OTU_SF_merge_site)
```

####STEP14.3:Filter extracted OTU table 
Extracted OTU table will be filtered for sum of row which is equal to 1 in 1 column
```{r}
OTU_SF_merge_site_endemic = OTU_SF_merge_site[rowSums(OTU_SF_merge_site == 0) == 3, ]
head(OTU_SF_merge_site_endemic)
```

####STEP14.4:Add OTUs to data frame
Add OTUs which are endemics to data frame which will be use in the next step for OTU selection from full dataset.
```{r}
OTU_SF_merge_site_endemic_df = data.frame(OTUendemic = row.names(OTU_SF_merge_site_endemic))
dim(OTU_SF_merge_site_endemic_df)
```

###STEP15: Import OTU table and filtering
Import full data OTU table again, then filter/select OTUs which are endemic species from data frame created in STEP14.3

```{r warning=FALSE}
otu <- read.table("MC2018FC.otu_table.txt",header=T,sep="\t")
otus.filtered = otu[otu$OTU.ID %in% OTU_SF_merge_site_endemic_df$OTUendemic,]
otus.filtered.rn = otus.filtered[,-1]
rownames(otus.filtered.rn) = otus.filtered[,1]
otumat.endemic <- as(as.matrix(otus.filtered.rn), "matrix")
OTU.endemic = otu_table(otumat.endemic, taxa_are_rows = TRUE)
```

###STEP16:Construct Phyloseq object for endemic species
```{r warning=FALSE}
physeq.endemic = phyloseq(OTU.endemic,TAX,sampleData,tree)
```

```{r}
physeq.endemic
```

###STEP17:Remove singletons
Remove any OTUs that present only one time.
```{r}
physeq.endemic.prune = prune_taxa(taxa_sums(physeq.endemic) > 1, physeq.endemic)
```

```{r}
physeq.endemic.prune
```

###STEP18: Plot read counts to check dataset
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

TotalReads of all the samples can be in this table (select only SampleID and TotalReads columns).
In order to check samples with low number of reads, "order()" can be used to sort "TotalReads" column.
In this dataset, N55.Rhizo has very low number of reads, so will will filter this sample out using the next minimum number of reads.
```{r}
readcount = data.table(as(sample_data(physeq.endemic.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.endemic.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
head(readcount)
```

###STEP19: Rarefy OTUs to a minimum number of reads
Rarefy OTUs (remove any samples that has very low number of reads)
If this step work fine, you have successfully imported data to R and completely generate phyloseq object as "physeq.endemic.prune.rarefy".

```{r warning=FALSE}
set.seed(711)
physeq.endemic.prune.rarefy = rarefy_even_depth(physeq.endemic.prune, sample.size = 192, replace = FALSE, trimOTUs = TRUE)
physeq.endemic.prune.rarefy
```


###STEP19.1:Remove non-Fungi OTUs
```{r}
physeq.endemic.prune.rarefy = subset_taxa(physeq.endemic.prune.rarefy, Kingdom != "Rhizaria")
physeq.endemic.prune.rarefy = subset_taxa(physeq.endemic.prune.rarefy, Kingdom != "Unassigned")
```

###STEP20:Merge samples by metadata for selected endemic species
For endemic analysis, we would like to look for endemic species at each sites. Therefore, we will merge samples from each site together. For examples, all of samples from GMT will be merged into one group as GMT. After merging, there will be 4 groups which are GMT, KELSO, CIMA, and JTNP.

```{r}
physeq.endemic.prune.rarefy.merge.sitelayer <- merge_samples(physeq.endemic.prune.rarefy, "Site")
sample_data(physeq.endemic.prune.rarefy.merge.sitelayer)$Site <- factor(sample_names(physeq.endemic.prune.rarefy.merge.sitelayer))
physeq.endemic.prune.rarefy.merge.sitelayer
```

###STEP21:Plot heatmap for all fungal endemic species
```{r}
Heatmap.sitelayer.endemic<- prune_taxa(names(sort(taxa_sums(physeq.endemic.prune.rarefy.merge.sitelayer),TRUE)[1:50]), physeq.endemic.prune.rarefy.merge.sitelayer)
Heatmap.sitelayer.endemic
```


```{r}
badTaxa = c("OTU58",  "OTU125",  "OTU254",  "OTU639")
goodTaxa <- setdiff(taxa_names(Heatmap.sitelayer.endemic), badTaxa)
Heatmap.sitelayer.endemic2 <- prune_taxa(goodTaxa, Heatmap.sitelayer.endemic)
Heatmap.sitelayer.endemic2
```


Results show 2 main phylum for endemic species which include Ascomycota and Basidiomycota
```{r}
Heatmap.sitelayer.endemic.plot = plot_heatmap(Heatmap.sitelayer.endemic2, "PCoA", "bray", "Site", high="orangered3")+ ggtitle("Endemic Fungal heatmap (top 50 OTUs)") + theme(plot.title = element_text(hjust = 0.5))
print(Heatmap.sitelayer.endemic.plot)

ggsave("./Figures/Fungal_endemic_species_heatmap_top50_Order_7Aug19.png", plot = Heatmap.sitelayer.endemic.plot, width = 4, height = 8)
```

```{r}
Heatmap.sitelayer.endemic2.order_taxglom = tax_glom(Heatmap.sitelayer.endemic2, taxrank = "Order")
#Heatmap.sitelayer.endemic2.order_taxglom.plot = plot_heatmap(Heatmap.sitelayer.endemic2.order_taxglom, "PCoA", "bray", "Site_layer", "Order")+ ggtitle("Fungal endemic species heatmap (top 50 OTUs taxglom order)") + theme(plot.title = element_text(hjust = 0.5))
#print(Heatmap.sitelayer.endemic2.order_taxglom.plot)
```

```{r}
tax_table(Heatmap.sitelayer.endemic2.order_taxglom)
```

```{r}
tax_table(Heatmap.sitelayer.endemic)
```


###STEP22.1:Plot heatmap for Ascomycota endemic species
```{r}
Heatmap.sitelayer.endemic.asco = subset_taxa(physeq.endemic.prune.rarefy.merge.sitelayer, Phylum == "Ascomycota")
Heatmap.sitelayer.endemic.asco <- prune_taxa(names(sort(taxa_sums(Heatmap.sitelayer.endemic.asco),TRUE)[1:50]), Heatmap.sitelayer.endemic.asco)
Heatmap.sitelayer.endemic.asco.plot = plot_heatmap(Heatmap.sitelayer.endemic.asco, "PCoA", "bray", "Site_layer", "Order")+ ggtitle("Endemic Ascomycota heatmap") + theme(plot.title = element_text(hjust = 0.5))
print(Heatmap.sitelayer.endemic.asco.plot)

ggsave("/Volumes/GoogleDrive/My Drive/UCR2019/Mojave_project2019/MojaveCrusts_2019analysis/EndemicSpecies/Fungi/Figures/Endemic Ascomycota Heatmap.png", plot = Heatmap.sitelayer.endemic.asco.plot, width = 12, height = 8)

#Mac's preview doesn't work well with heatmap. Still not sure how to fix this but will use png for now.
#pdf("/Volumes/GoogleDrive/My Drive/UCR2018/MacAir2018/Mojave_Project/Mojave_16S_Qiime2/Figures/Bacterial Heatmap by Site and Layer.pdf" )
#Heatmap.sitelayer.cyano.plot
#dev.off()
```

```{r}
tax_table(Heatmap.sitelayer.endemic.asco)
```


###STEP22.2:Plot tree for Ascomycota endemic species
```{r}
plot_tree(Heatmap.sitelayer.endemic.asco, ladderize="left", color="Site_layer", label.tips = "Order", text.size = 2) + ggtitle("Endemic Ascomycota tree") + theme(plot.title = element_text(hjust = 0.5))
```




